tracer/src/Datadog.Trace/Propagators/W3CBaggagePropagator.cs

```diff
@@ -0,0 +1,345 @@
+    {
+        var baggage = context.Baggage;
+
+        if (baggage is null or { Count: 0 })
+        {
+            // nothing to inject
+            return;
+        }
+
+        var settings = Tracer.Instance.Settings;
+        var headerValue = CreateHeader(baggage, settings.BaggageMaximumItems, settings.BaggageMaximumBytes);
+
+        if (!string.IsNullOrWhiteSpace(headerValue))
+        {
+            carrierSetter.Set(carrier, BaggageHeaderName, headerValue);
+            TelemetryFactory.Metrics.RecordCountContextHeaderStyleInjected(MetricTags.ContextHeaderStyle.Baggage);
+        }
+    }
+
+    public bool TryExtract<TCarrier, TCarrierGetter>(
+        TCarrier carrier,
+        TCarrierGetter carrierGetter,
+        out PropagationContext context)
+        where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+    {
+        context = default;
+        var header = ParseUtility.ParseString(carrier, carrierGetter, BaggageHeaderName);
+
+        if (string.IsNullOrWhiteSpace(header))
+        {
+            return false;
+        }
+
+        var baggage = ParseHeader(header!);
+        context = new PropagationContext(spanContext: null, baggage);
+        TelemetryFactory.Metrics.RecordCountContextHeaderStyleExtracted(MetricTags.ContextHeaderStyle.Baggage);
+
+        return baggage is { Count: > 0 };
+    }
+
+    internal static string Encode(string source, HashSet<char> charsToEncode)
+    {
+        if (source.Length == 0)
+        {
+            return string.Empty;
+        }
+
+        // this is an upper bound and will almost always be more bytes than we need
+        var maxByteCount = Encoding.UTF8.GetMaxByteCount(source.Length);
+
+#if NETCOREAPP3_1_OR_GREATER
+        if (maxByteCount < 256)
+        {
+            return EncodeOnStack(maxByteCount, source, charsToEncode);
+
+            static string EncodeOnStack(int maxByteCount, string source, HashSet<char> charsToEncode)
+            {
+                // allocate a buffer on the stack for the UTF-8 bytes
+                Span<byte> buffer = stackalloc byte[maxByteCount];
+                var byteCount = Encoding.UTF8.GetBytes(source, buffer);
+
+                // slice the buffer down to the actual bytes written
+                var bytes = buffer[..byteCount];
+                return EncodeBytes(source, bytes, charsToEncode);
+            }
+        }
+#endif
+
+        // rent a buffer for the UTF-8 bytes
+        var buffer = ArrayPool<byte>.Shared.Rent(maxByteCount);
+        var byteCount = Encoding.UTF8.GetBytes(source, 0, source.Length, buffer, 0);
+
+        // slice the buffer down to the actual bytes written
+        var bytes = buffer.AsSpan(0, byteCount);
+        var result = EncodeBytes(source, bytes, charsToEncode);
+
+        ArrayPool<byte>.Shared.Return(buffer);
+        return result;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        static bool ByteRequiresEncoding(byte b, HashSet<char> charsToEncode)
```

Comment:
I wish we had all the `SearchValues` stuff from .NET 8 available, that would do all this sort of thing for us 😢

Reply:
TIL about `SearchValues`